subsys/net/lib/websocket/websocket.c

```diff
@@ -0,0 +1,1529 @@
+
+	memset(header, 0, sizeof(header));
+
+	/* Is this the last packet? */
+	header[0] = final ? BIT(7) : 0;
+
+	/* Text, binary, ping, pong or close ? */
+	header[0] |= opcode;
+
+	/* Masking */
+	header[1] = mask ? BIT(7) : 0;
+
+	if (len < 126) {
+		header[1] |= len;
+	} else if (len == 126) {
+		header[1] |= 126;
+		header[2] |= htons(len) >> 8;
+		header[3] |= htons(len);
+		hdr_len += 2;
+	} else {
+		header[1] |= 127;
+		header[2] |= htonl(len) >> 24;
+		header[3] |= htonl(len) >> 16;
+		header[4] |= htonl(len) >> 8;
+		header[5] |= htonl(len);
+		hdr_len += 4;
+	}
+
+	/* Add masking value if needed */
+	if (mask) {
+		u32_t masking_value;
+
+		masking_value = sys_rand32_get();
+
+		header[hdr_len++] |= masking_value >> 24;
+		header[hdr_len++] |= masking_value >> 16;
+		header[hdr_len++] |= masking_value >> 8;
+		header[hdr_len++] |= masking_value;
+
+		mask_pkt(pkt, masking_value);
+	}
+
+	if (!net_pkt_insert(pkt, pkt->frags, 0, hdr_len, header,
+			    ctx->timeout)) {
+		return -ENOMEM;
+	}
+
+	net_pkt_set_appdatalen(pkt, net_pkt_appdatalen(pkt) + hdr_len);
+
+	ret = ws_send_data_raw(ctx, pkt, user_send_data);
+	if (ret < 0) {
+		NET_DBG("Cannot send %zd bytes message (%d)",
+			net_pkt_get_len(pkt), ret);
+	}
+
+	return ret;
+}
+
+int ws_add_http_header(struct ws_ctx *ctx, struct net_pkt **pkt,
+		       const char *http_header)
+{
+	if (!pkt) {
+		return -EINVAL;
+	}
+
+	if (!*pkt) {
+		*pkt = net_app_get_net_pkt(&ctx->app_ctx, AF_UNSPEC,
+					   ctx->timeout);
+		if (!*pkt) {
+			return -ENOMEM;
+		}
+	}
+
+	return _ws_add_header(*pkt, ctx->timeout, http_header);
+}
+
+int ws_add_http_data(struct ws_ctx *ctx, struct net_pkt **pkt,
+		     const u8_t *buf, size_t len)
+{
+	if (!pkt) {
+		return -EINVAL;
+	}
+
+	if (!*pkt) {
+		*pkt = net_app_get_net_pkt(&ctx->app_ctx, AF_UNSPEC,
+					   ctx->timeout);
+		if (!*pkt) {
+			return -ENOMEM;
+		}
+	}
+
+	return _ws_add_chunk(*pkt, ctx->timeout, buf, len);
+}
+
+#if defined(CONFIG_NET_DEBUG_WS) && (CONFIG_SYS_LOG_NET_LEVEL > 2)
+static char *sprint_ipaddr(char *buf, int buflen, const struct sockaddr *addr)
+{
+	if (addr->sa_family == AF_INET6) {
+#if defined(CONFIG_NET_IPV6)
+		char ipaddr[NET_IPV6_ADDR_LEN];
+
+		net_addr_ntop(addr->sa_family,
+			      &net_sin6(addr)->sin6_addr,
+			      ipaddr, sizeof(ipaddr));
+		snprintk(buf, buflen, "[%s]:%u", ipaddr,
+			 ntohs(net_sin6(addr)->sin6_port));
+#endif
+	} else if (addr->sa_family == AF_INET) {
+#if defined(CONFIG_NET_IPV4)
+		char ipaddr[NET_IPV4_ADDR_LEN];
+
+		net_addr_ntop(addr->sa_family,
+			      &net_sin(addr)->sin_addr,
+			      ipaddr, sizeof(ipaddr));
+		snprintk(buf, buflen, "%s:%u", ipaddr,
+			 ntohs(net_sin(addr)->sin_port));
+#endif
+	}
+
+	return buf;
+}
+#endif /* CONFIG_NET_DEBUG_HTTP */
+
+static inline void new_client(struct ws_ctx *ctx,
+			      enum ws_connection_type type,
+			      struct net_app_ctx *app_ctx)
+{
+#if defined(CONFIG_NET_DEBUG_WS) && (CONFIG_SYS_LOG_NET_LEVEL > 2)
```

Comment:
This block seems to be duplicated quite a lot.  Can we #define PORT_STR at the top?   Also, PORT_STR seems to indicate it's a string.  Not a sizeof a string.

Reply:
I will refactor this